### PR TITLE
Make ProvidedMethod invokeCode parameter optional #180

### LINKED ISSUE #180: mandatory invokeCode parameter in ProvidedMethod ctor senseless for statically parametrized TP methods
### Description

A new ProvidedTypes api is great but I wonder what is supposed to be passed as value for ```invokeCode``` https://github.com/fsprojects/FSharp.TypeProviders.SDK/blob/33fd1c0a95dd1ddfe77f12e4921b6983f72c8e62/src/ProvidedTypes.fsi#L96 parameter for  [statically parametrized TP methods](https://github.com/fsharp/fslang-design/blob/master/FSharp-4.0/StaticMethodArgumentsDesignAndSpec.md).

### Known workarounds

For now I used bogus value ```Unchecked.defaultof<_>```
```fsharp
    let m = ProvidedMethod("CreateCommand", [], typeof<obj>, isStatic = true, invokeCode = Unchecked.defaultof<_>)
    m.DefineStaticParameters(staticParams, (fun methodName args ->
    ... 
```

### PR DESCRIPTION
**Pull request by panesofglass**
This is an attempt to resolve #180 by making `invokeCode` an optional parameter on `ProvidedMethod`.

### CODE REVIEW

File: src/ProvidedTypes.fs [hunk 1]
```diff
@@ -14365,8 +14368,11 @@ namespace ProviderImplementation.ProvidedTypes
 
                 match methodBaseT with
                 | :? ProvidedMethod as mT when (match methodBaseT.DeclaringType with :? ProvidedTypeDefinition as pt -> pt.IsErased | _ -> true) ->
-                    let exprT = mT.GetInvokeCode(Array.toList parametersT)
-                    check exprT
+                    match mT.GetInvokeCode with
+                    | Some invokeCode ->
+                        let exprT = invokeCode(Array.toList parametersT)
+                        check exprT
+                    | None -> <@@ () @@>
```
Comment (panesofglass) [inline src/ProvidedTypes.fs]: I wasn't sure what to do here. @dmitry-a-morozov suggested `Unchecked.defaultof<_>`, but generative providers appear to fail due to a null reference. @7sharp9 suggested this solution, so I've made a go with it as a first pass.

@dmitry-a-morozov, would you be able to add some test cases?

Comment (dmitry-a-morozov) [reply src/ProvidedTypes.fs]: @panesofglass I'm sorry but no. I don't have capacity to work on it now.

Comment (dsyme) [inline src/ProvidedTypes.fs]: We should raise an exception at this point if a body has not been provided where one will be needed (i.e. when the method is not abstract) - indeed earlier, in the ProvidedMethod constructor.  Likewise we should raise an exception if a body has been provided for a method that can't have one (an interface instance member)

**Comment by dsyme** [comment]
@panesofglass This is basically good but we should do the following checking in the ProvidedMethod/ProvidedConstructor/ProvidedProperty construction process

1. fail if the method is an interface instance method and has code
1. fail if the method is a non-abstract method and does not have code

In both cases give a good error message saying that the code parameter is/isn't not allowed.

Add unit tests for these in the test project.

**Comment by dsyme** [comment]
@panesofglass I'll aim to take a look tomorrow

File: src/ProvidedTypes.fs [hunk 2]
```diff
@@ -14365,12 +14376,24 @@ namespace ProviderImplementation.ProvidedTypes
...
+                    | Some _ when methodBaseT.DeclaringType.IsInterface ->
+                        failwith "The provided type definition is an interface; therefore, it should not define an implementation for its members."
+                    | Some _ when mT.IsAbstract ->
+                        failwith "The provided method is defined as abstract; therefore, it should not define an implementation."
+                    | None when not mT.IsAbstract ->
+                        failwith "The provided method is not defined as abstract; therefore it should define an implementation."
+                    | Some invokeCode ->
+                        let exprT = invokeCode(Array.toList parametersT)
+                        check exprT
+                    | None -> <@@ () @@>
```
Comment (panesofglass) [inline src/ProvidedTypes.fs]: I'm not sure these checks make sense for erased types, but perhaps they are?

File: src/ProvidedTypes.fs [hunk 3]
```diff
@@ -14365,12 +14376,24 @@ namespace ProviderImplementation.ProvidedTypes
...
 
                 | :? ProvidedConstructor as mT when (match methodBaseT.DeclaringType with :? ProvidedTypeDefinition as pt -> pt.IsErased | _ -> true) ->
-                    let exprT = mT.GetInvokeCode(Array.toList parametersT)
-                    check exprT
+                    if methodBaseT.DeclaringType.IsInterface then
+                        failwith "The provided type definition is an interface; therefore, it should not define any constructors."
+                    else
```
Comment (panesofglass) [inline src/ProvidedTypes.fs]: Is this check for erased constructors useful?

File: src/ProvidedTypes.fs [hunk 4]
```diff
@@ -14126,13 +14128,22 @@ namespace ProviderImplementation.ProvidedTypes
...
-                        let expectedState = if (transType minfo.ReturnType = ILType.Void) then ExpectedStackState.Empty else ExpectedStackState.Value
-                        let codeGen = CodeGenerator(assemblyMainModule, genUniqueTypeName, implicitCtorArgsAsFields, convTypeToTgt, transType, transFieldSpec, transMeth, transMethRef, transCtorSpec, ilg, methLocals, parameterVars)
-                        codeGen.EmitExpr (expectedState, expr)
+                        match pminfo.GetInvokeCode with
+                        | Some _ when ptdT.IsInterface ->
+                            failwith "The provided type definition is an interface; therefore, it should not define an implementation for its members."
+                        | Some _ when pminfo.IsAbstract ->
+                            failwith "The provided method is marked as an abstract method; therefore, it should not define an implementation."
+                        | None when not pminfo.IsAbstract ->
+                            failwith "The provided method is not marked as an abstract method; therefore, it should define an implementation."
```
Comment (panesofglass) [inline src/ProvidedTypes.fs]: I added checks for provided methods.

File: src/ProvidedTypes.fs [hunk 5]
```diff
@@ -14097,6 +14097,8 @@ namespace ProviderImplementation.ProvidedTypes
 
                     match ptdT.GetConstructors(bindAll) |> Seq.tryPick (function :? ProvidedConstructor as pc when pc.IsTypeInitializer -> Some pc | _ -> None) with
                     | None -> ()
+                    | Some _ when ptdT.IsInterface ->
+                        failwith "The provided type definition is an interface; therefore, it may not provide constructors."
```
Comment (panesofglass) [inline src/ProvidedTypes.fs]: I added this one check for provided constructors for interfaces. Abstract classes can still have constructors, so I left it to interfaces alone.

**Comment by panesofglass** [comment]
I'm amazed I got within one line!

**Comment by panesofglass** [comment]
Will this be merged soon?

**Comment by dsyme** [comment]
Thanks for this!

**Pull request merged**